### PR TITLE
Feature/sample history

### PR DESCRIPTION
**Pull request by halfhorst**

### CODE REVIEW

File: src/vivarium_conic_sqlns/components/observers.py [hunk 1]
```diff
@@ -9,3 +13,91 @@ def setup(self, builder):
...
         self.disability_weight_pipelines['iron_deficiency'] = builder.value.get_value('iron_deficiency.disability_weight')
+
+
+class SampleHistoryObserver:
+
+    configuration_defaults = {
+        'metrics': {
+            'sample_history': {
+                'sample_size': 100,
+                'path': '/dev/null'
```
Comment (collijk) [inline src/vivarium_conic_sqlns/components/observers.py]: Use `~/sample_history.hdf` as your default.

File: src/vivarium_conic_sqlns/components/observers.py [hunk 2]
```diff
@@ -9,3 +13,91 @@ def setup(self, builder):
...
+        config = builder.configuration['metrics']['sample_history']
+        self.sample_size = config['sample_size']
+        self.path = pathlib.Path(config['path']).resolve()
+        if self.path.suffix != '.hdf':
+            raise ValueError("metrics: sample_history: path must specify a path to an HDF file.")
+
+        self.clock = builder.time.clock()
+        self.randomness = builder.randomness.get_stream('sample_index')
+
+        self.builder = builder
```
Comment (collijk) [inline src/vivarium_conic_sqlns/components/observers.py]: pull

File: src/vivarium_conic_sqlns/components/observers.py [hunk 3]
```diff
@@ -9,3 +13,91 @@ def setup(self, builder):
...
+        self.population_view = builder.population.get_view(['alive', 'age', 'sex', 'exit_time'])
```
Comment (collijk) [inline src/vivarium_conic_sqlns/components/observers.py]: treatment start and end

File: src/vivarium_conic_sqlns/components/observers.py [hunk 4]
```diff
@@ -9,3 +13,91 @@ def setup(self, builder):
...
+
+        self.pipelines = {
+            'iron_deficiency_exposure': builder.value.get_value('iron_deficiency.exposure'),
+            'child_stunting_exposure': builder.value.get_value('child_stunting.exposure'),
+            'child_wasting_exposure': builder.value.get_value('child_wasting.exposure'),
+            'lower_resipratory_infections_incidence_rate': builder.value.get_value('lower_respiratory_infections.incidence_rate'),
+            'diarrheal_diseases_incidence_rate' : builder.value.get_value('diarrheal_diseases.incidence_rate'),
+            'measles_incidence_rate': builder.value.get_value('measles.incidence_rate'),
+            'disability_weight': builder.value.get_value('disability_weight')
```
Comment (collijk) [inline src/vivarium_conic_sqlns/components/observers.py:62]: add cause specific disability weights

File: src/vivarium_conic_sqlns/components/observers.py [hunk 5]
```diff
@@ -9,3 +13,91 @@ def setup(self, builder):
...
+        }
+
+        builder.population.initializes_simulants(self.get_sample_index)
+        builder.event.register_listener('collect_metrics', self.record)
+        builder.event.register_listener('simulation_end', self.dump_history)
+
+    def get_sample_index(self, pop_data):
+        if self.sample_index is None:
```
Comment (collijk) [inline src/vivarium_conic_sqlns/components/observers.py]: Make a sample proportion, use filter for probability and append to your sampled index.

File: src/vivarium_conic_sqlns/components/observers.py [hunk 6]
```diff
@@ -9,3 +13,91 @@ def setup(self, builder):
...
+
+            draw = self.randomness.get_draw(pop_data.index)
+            priority_index = [i for d, i in sorted(zip(draw, pop_data.index), key=lambda x:x[0])]
+            self.sample_index = pd.Index(priority_index[:self.sample_size])
+
+    def record(self, event):
+        pop = self.population_view.get(self.sample_index)
+        pipeline_results = []
+        for name, pipeline in self.pipelines.items():
+            if name == 'iron_deficiency':
```
Comment (collijk) [inline src/vivarium_conic_sqlns/components/observers.py]: iron_deficiency_exposure
wasting, stunting

File: src/vivarium_conic_sqlns/components/observers.py [hunk 7]
```diff
@@ -9,3 +13,91 @@ def setup(self, builder):
...
+                skip_post_processor = False
+
+            # Using pop.index due to untracked individuals
+            values = pipeline(pop.index, skip_post_processor=skip_post_processor)
+            values = values.rename(name)
+            pipeline_results.append(values)
+
+            raw_values = pipeline.source(pop.index)
+            if isinstance(raw_values, list):
+                raw_values = raw_values[0]
```
Comment (collijk) [inline src/vivarium_conic_sqlns/components/observers.py]: find out the name and let me know

Comment (collijk) [reply src/vivarium_conic_sqlns/components/observers.py]: nvm, disability weight has a list combiner and a joint value postprocessor, so a list for the source.

File: src/vivarium_conic_sqlns/components/observers.py [hunk 8]
```diff
@@ -9,3 +13,91 @@ def setup(self, builder):
...
+                skip_post_processor = True  # raw hemoglobin was requested
+            else:
+                skip_post_processor = False
+
+            # Using pop.index due to untracked individuals
+            values = pipeline(pop.index, skip_post_processor=skip_post_processor)
+            values = values.rename(name)
+            pipeline_results.append(values)
+
+            raw_values = pipeline.source(pop.index)
```
Comment (collijk) [inline src/vivarium_conic_sqlns/components/observers.py]: If name != disability weight

Comment (collijk) [reply src/vivarium_conic_sqlns/components/observers.py]: No disability weight pipeline sources, they're super confusing.

Comment (collijk) [reply src/vivarium_conic_sqlns/components/observers.py]: No incidence pipelines (Both SOURCES) either, pipeline.source is risk-deleted, not un-intervened.

**Review by collijk — CHANGES_REQUESTED** [review]

File: src/vivarium_conic_sqlns/components/observers.py [hunk 9]
```diff
@@ -9,3 +13,88 @@ def setup(self, builder):
...
+            else:
+                skip_post_processor = False
+
+            # Using pop.index due to untracked individuals
+            values = pipeline(pop.index, skip_post_processor=skip_post_processor)
+            values = values.rename(name)
+            pipeline_results.append(values)
+
+            # some pipelines sources aren't really a 'baseline'
+            if name != 'disability_weight' and not name.endswith('_incidence_rate'):
```
Comment (halfhorst) [inline src/vivarium_conic_sqlns/components/observers.py]: removing the cause dw now, I misinterpreted

**Comment by collijk** [comment]
Add disability for iron deficiency and pem and then you're good to go.

**Review by collijk — APPROVED** [review]

**Pull request merged**